stacks/dotnet31/devfile.yaml

```diff
@@ -0,0 +1,50 @@
+schemaVersion: 2.1.0
+metadata:
+  name: dotnet50
+  displayName: .NET 3.1
+  description: Stack with .NET 3.1
+  icon: https://developers.redhat.com/sites/default/files/styles/large/public/dotnet-logo3x3.jpg
+  projectType: dotnet
+  version: 1.0.0
+
+starterProjects:
+  - name: dotnetcore
+    git:
+      checkoutFrom:
+        remote: origin
+        revision: dotnetcore-3.1
+      remotes:
+        origin: https://github.com/redhat-developer/s2i-dotnetcore-ex
+    subDir: app
```

Comment:
What does subDir do, does it trim the repo to this folder?

Often .NET repos have sibling folders next to the app that contain additional code that gets used by the app.

Is this pattern supported by `subDir`?

Reply:
We should remove `subdir` here, and replace it with a parameter on the component.

The parameter can be called `STARTUP_PROJECT`/`PROJECT`.

In our s2i builder it is called `DOTNET_STARTUP_PROJECT` (see https://github.com/redhat-developer/s2i-dotnetcore/blob/master/5.0/build/README.md#environment-variables).

Reply:
But the application source code is in the `app` forlder no? In the root of the repo there is just documentation.

If I add something like `DOTNET_STARTUP_PROJECT` I will have to set its default value to `app` to make the starter project work out of the box. And this will mean that every time someone will want to use this devfile with a project where the application in root of the repository they will have to modify `DOTNET_STARTUP_PROJECT`

Reply:
> But the application source code is in the app forlder no? In the root of the repo there is just documentation.

Yes, for `s2i-dotnetcore-ex` it is. This is basic `hello` world web app. It doesn't have library projects.

> If I add something like DOTNET_STARTUP_PROJECT I will have to set its default value to app to make the starter project work out of the box. And this will mean that every time someone will want to use this devfile with a project where the application in root of the repository they will have to modify DOTNET_STARTUP_PROJECT

.NET apps rarely live in the repository root, so something needs to be set anyhow. Instead of `subdir`, the user sets `STARTUP_PROJECT`.

Reply:
I'm going to make some time still this week to get some first hand experience using `odo` with `Devfiles`.

Reply:
@tmds I've added a bit of "bash magic" into the commands, so if the `STARTUP_PROJECT` is not set, it will automatically pick first *.csfile or *.fsfile. Not sure if it is good idea or not. If you think that it might be confusing, I can remove it, and we can rely only on `STARTUP_PROJECT`

Reply:
> I've added a bit of "bash magic" into the commands, so if the STARTUP_PROJECT is not set, it will automatically pick first *.csfile or *.fsfile. Not sure if it is good idea or not. If you think that it might be confusing, I can remove it, and we can rely only on STARTUP_PROJECT

Let's remove it. The user can easily figure out how to use `STARTUP_PROJECT`. Not all users will grok the bash magic.